docs/develop/smart-contracts/guidelines/message-delivery-guarantees.mdx

```diff
@@ -0,0 +1,70 @@
+
+<ConceptImage src="/img/docs/msg-delivery-1.png" />
+
+It is just a simple case when we have only two contracts. What if our system works with more of them?
+
+### Several smart contracts
+
+Suppose that we have three contracts - _A_, _B_ and _C_. In some transaction, _A_ sends two internal messages _1_ and _2_: one to _B_ and another one to _C_. Even though they were created in some exact order (_1_, then _2_), we can't be sure that _1_ will be processed before _2_. That happens because routes from _A_ to _B_ and from _A_ to _C_ can differ in length and validator sets. If these contracts are in different shardchains, one of the messages may require several blocks to reach the destination contract.
+
+<ConceptImage src="/img/docs/msg-delivery-2.png" />
+
+The same thing happens in a reverse situation, when two contracts _B_ and _C_ send a message to one contract _A_. Even if message `B -> A` was sent before `C -> A`, we can't know which of them will be deliveried first. `B -> A` route may require more shardchain hops.
+
+<ConceptImage src="/img/docs/msg-delivery-3.png" />
+
+### Chains of messages
+
+Even more interesting things can happend when we have chains of messages. Let's have a look at a few examples.
+
+Suppose that we have contracts A, B and C. In some transcation, A sends two internal messages 1 and 2: one to B and one to C. And after receiving 1, contract B also sends a message 3 to C. In this case we can be 100% sure that message 2 will be deliveried before message 3, because the path `A -> C` is shorter than `A -> B -> C`.
+
+<ConceptImage src="/img/docs/msg-delivery-4.png" />
+
+Next example is similar to what was described in [Several smart contracts](#several-smart-contracts). Let's say that we have contracts A, B, C and D. In some transaction, A sends an internal message 1 to B and an internal message 2 to C. Both B and C will send a message (3 and 4) to D after receiving a message from A. In this case, we also can't be sure which message, 3 or 4 will be deliveried first.
+
+<ConceptImage src="/img/docs/msg-delivery-5.png" />
+
+And the most absurd example is shown on the image below. In this case we also can't say which message will be deliveried to B first. Yes, the chances of message 2 being faster than 6 are obviously higher, but still not 100%.
+
+<ConceptImage src="/img/docs/msg-delivery-6.png" />
+
+## Conclusion
+
+TON is very fast and scalable blockchain, but the price for it is that developer should be more careful and accurate when developing smart contracts. If the logic of the contract depends on the order in which messages are processed, the developer needs to carefully examine the guarantees that TON provides and make sure that the desired order is maintained even when the contracts are in different shards or workchains.
```

Comment:
TON is **a** very fast and scalable blockchain, but the price for it is that **a** developer **must** be more careful and accurate when **creating** smart contracts. If the logic of the contract depends on the order in which messages are processed, the developer needs to carefully examine the guarantees that TON provides and make sure that the desired order is maintained even when the contracts are in different shards or workchains.